drm-howto/modeset-atomic.c

```diff
@@ -0,0 +1,1181 @@
+ * displaying unwanted content.
+ *
+ * NOTE: we can't perform an atomic commit without an attached frambeuffer
+ * (even when we have DRM_MODE_ATOMIC_TEST_ONLY). It will simply fail.
+ */
+
+static int modeset_perform_modeset(int fd)
+{
+	int ret, flags;
+	struct modeset_output *iter;
+	drmModeAtomicReq *req;
+
+	/* prepare modeset on all outputs */
+	req = drmModeAtomicAlloc();
+	for (iter = output_list; iter; iter = iter->next) {
+		ret = modeset_atomic_prepare_commit(fd, iter, req);
+		if (ret < 0)
+			break;
+	}
+	if (ret < 0) {
+		fprintf(stderr, "prepare atomic commit failed, %d\n", errno);
+		return ret;
+	}
+
+	/* perform test-only atomic commit */
+	flags = DRM_MODE_ATOMIC_TEST_ONLY | DRM_MODE_ATOMIC_ALLOW_MODESET;
+	ret = drmModeAtomicCommit(fd, req, flags, NULL);
+	if (ret < 0) {
+		fprintf(stderr, "test-only atomic commit failed, %d\n", errno);
+		drmModeAtomicFree(req);
+		return ret;
+	}
+
+	/* draw on back framebuffer of all outputs */
+	for (iter = output_list; iter; iter = iter->next) {
+
+		/* colors initialization, this is the first time we're drawing */
+		iter->r = rand() % 0xff;
+		iter->g = rand() % 0xff;
+		iter->b = rand() % 0xff;
+		iter->r_up = iter->g_up = iter->b_up = true;
+
+		modeset_paint_framebuffer(iter);
+	}
+
+	/* initial modeset on all outputs */
+	flags = DRM_MODE_ATOMIC_ALLOW_MODESET | DRM_MODE_PAGE_FLIP_EVENT;
+	ret = drmModeAtomicCommit(fd, req, flags, NULL);
+	if (ret < 0)
+		fprintf(stderr, "modeset atomic commit failed, %d\n", errno);
+
+	drmModeAtomicFree(req);
+
+	return ret;
+}
+
+/*
+ * modeset_draw() changes. If we got here, the modeset already occurred. When
+ * the page-flip for a certain output is done, an event will be fired and we'll
+ * be able to handle it.
+ *
+ * Here we define the function that should handle these events, which is
+ * modeset_page_flip_event(). This function calls modeset_draw_out(), which is
+ * responsible for preparing a new framebuffer and performing another atomic
+ * commit for us.
+ *
+ * Then we have a 5 seconds loop that keeps waiting for the events that are
+ * fired when the page-flip is complete. drmHandleEvent() is reponsible for
+ * reading the events from the fd and to call modeset_page_flip_event() for
+ * each one of them.
+ */
+
+static void modeset_draw(int fd)
+{
+	int ret;
+	fd_set fds;
+	time_t start, cur;
+	struct timeval v;
+	drmEventContext ev;
+
+	/* init variables */
+	srand(time(&start));
+	FD_ZERO(&fds);
+	memset(&v, 0, sizeof(v));
+	memset(&ev, 0, sizeof(ev));
+
+	/* 3 is the first version that allow us to use page_flip_handler2, which
+	 * is just like page_flip_handler but with the addition of passing the
+	 * crtc_id as argument to the function that will handle page-flip events
+	 * (in our case, modeset_page_flip_event()). This is good because we can
+	 * find out for what output the page-flip happened.
+	 */
+	ev.version = 3;
+	ev.page_flip_handler2 = modeset_page_flip_event;
```

Comment:
This requires checking `drmGetCap(b->drm.fd, DRM_CAP_CRTC_IN_VBLANK_EVENT, &cap);` like Weston would put it.

Reply:
Oh sorry, you [had already explained this](https://github.com/dvdhrm/docs/pull/10#discussion_r408893910) and I forgot to add the check. Fixed.

Reply:
Could still mention in the comment here that this is what we needed `DRM_CAP_CRTC_IN_VBLANK_EVENT` for.

Reply:
Done.